Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -14,21 +14,183 @@
+    private var config: KafkaConfig
     private let topicConfig: KafkaTopicConfig
+    private let logger: Logger
+    private var topicHandles: [String: OpaquePointer]
 
-    // Preliminary implementation
+    // Implicitly unwrapped optional because we must pass self to our config before initializing client etc.
+    private var referenceToSelf: UnsafeMutableRawPointer!
+    private var client: KafkaClient!
+    private var kafkaHandle: OpaquePointer!
+    private var pollTask: Task<Void, Never>!
+
+    /// Initialize a new `KafkaProducer`.
+    /// - Parameter config: The ``KafkaConfig`` for configuring the `KafkaProducer`.
+    /// - Parameter topicConfig: The ``KafkaTopicConfig`` used for newly created topics.
+    /// - Parameter logger: A logger.
     public init(
         config: KafkaConfig = KafkaConfig(),
         topicConfig: KafkaTopicConfig = KafkaTopicConfig(),
         logger: Logger
-    ) throws {
-        self.client = try KafkaClient(type: .producer, config: config, logger: logger)
+    ) async throws {
+        self.config = config
         self.topicConfig = topicConfig
+        self.logger = logger
+        self.topicHandles = [:]
+
+        self.referenceToSelf = Unmanaged.passRetained(self).toOpaque()
+        self.config.setDeliveryReportCallback(callback: self.deliveryReportCallback)
+        self.config.setCallbackOpaque(opaque: self.referenceToSelf)
+
+        self.client = try KafkaClient(type: .producer, config: self.config, logger: self.logger)
+        self.kafkaHandle = self.client.kafkaHandle
+
+        // Poll Kafka every millisecond
+        self.pollTask = Task { [kafkaHandle] in
+            while !Task.isCancelled {
+                rd_kafka_poll(kafkaHandle, 0)
+                try? await Task.sleep(nanoseconds: 1_000_000)
+            }
+        }
+    }
+
+    /// Kafka producer must be closed **explicitly**.
+    public func shutdownGracefully() {
+        // Wait 10 seconds for outstanding messages to be sent and callbacks to be called
+        rd_kafka_flush(self.kafkaHandle, 10000)
+
+        for (_, topicHandle) in self.topicHandles {
+            rd_kafka_topic_destroy(topicHandle)
+        }
+        self.pollTask.cancel()
+
+        _ = Unmanaged<KafkaProducer>.fromOpaque(self.referenceToSelf).takeRetainedValue()
+    }
+
+    /// Send messages to the Kafka cluster asynchronously, aka "fire and forget".
+    /// This function is non-blocking.
+    /// - Parameter message: The ``KafkaProducerMessage`` that is sent to the KafkaCluster.
+    /// - Returns: Unique message identifier matching the `id` property of the corresponding ``KafkaAckedMessage``
+    @discardableResult
+    public func sendAsync(message: KafkaProducerMessage) throws -> UInt {
+        let topicHandle = self.createTopicHandleIfNeeded(topic: message.topic)
+
+        let keyBytes: [UInt8]?
+        if let key = message.key {
+            keyBytes = key.getBytes(at: 0, length: key.readableBytes)
```

Comment:
Using the `getX` methods on `ByteBuffer` is almost never correct since it does not respect the current `readerIndex` and doesn't move it forward. Please use the `readBytes` method instead.  This will require you to create a local copy of the message which is okay. In the future, we could mark the `message` parameter as `taking` to indicate that it takes ownership of the value